File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/business_result/BusinessPaymentResultMapper.java
package com.mercadopago.android.px.internal.features.business_result;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.mercadopago.android.px.R;
import com.mercadopago.android.px.internal.features.dummy_result.RedirectHelper;
import com.mercadopago.android.px.internal.features.payment_congrats.model.CongratsViewModelMapper;
import com.mercadopago.android.px.internal.features.payment_congrats.model.PaymentCongratsModel;
import com.mercadopago.android.px.internal.features.payment_congrats.model.PaymentInfo;
import com.mercadopago.android.px.internal.view.PaymentResultBody;
import com.mercadopago.android.px.internal.view.PaymentResultHeader;
import com.mercadopago.android.px.internal.view.PaymentResultMethod;
import com.mercadopago.android.px.internal.viewmodel.GenericLocalized;
import com.mercadopago.android.px.internal.viewmodel.PaymentResultType;
import com.mercadopago.android.px.internal.viewmodel.mappers.Mapper;
import java.util.ArrayList;
import java.util.List;

public class BusinessPaymentResultMapper extends Mapper<PaymentCongratsModel, BusinessPaymentResultViewModel> {

    @Nullable private final String autoReturn;

    /* default */ BusinessPaymentResultMapper(@Nullable final String autoReturn) {
        this.autoReturn = autoReturn;
    }

    @Override
    public BusinessPaymentResultViewModel map(@NonNull final PaymentCongratsModel model) {
        final PaymentResultHeader.Model headerModel = getHeaderModel(model);
        final PaymentResultBody.Model bodyModel = getBodyModel(model);
        return new BusinessPaymentResultViewModel(headerModel, bodyModel,
            model.getFooterMainAction(), model.getFooterSecondaryAction(),
            RedirectHelper.INSTANCE.shouldAutoReturn(autoReturn, model.getCongratsType().name));
    }

    @NonNull
    private PaymentResultBody.Model getBodyModel(@NonNull final PaymentCongratsModel model) {
        final List<PaymentResultMethod.Model> methodModels = new ArrayList<>();
        if (model.getShouldShowPaymentMethod()) {
            for (final PaymentInfo paymentInfo : model.getPaymentsInfo()) {
                methodModels.add(PaymentResultMethod.Model.with(paymentInfo,
                    model.getStatementDescription()));
            }
        }
        final PaymentCongratsModel.CongratsType type = model.getCongratsType();
        return new PaymentResultBody.Model.Builder()
            .setMethodModels(methodModels)
            .setCongratsViewModel(new CongratsViewModelMapper(new BusinessPaymentResultTracker())
                .map(model.getPaymentCongratsResponse()))
            .setReceiptId((type == PaymentCongratsModel.CongratsType.APPROVED && model.getShouldShowReceipt()) ? String
                .valueOf(model.getPaymentId()) : null)
            .setHelp(model.getHelp())
            .setStatement(model.getStatementDescription())
            .setTopFragment(model.getTopFragment())
            .setBottomFragment(model.getBottomFragment())
            .setImportantFragment(model.getImportantFragment())
            .build();
    }

    @NonNull
    private PaymentResultHeader.Model getHeaderModel(@NonNull final PaymentCongratsModel model) {
        final PaymentResultHeader.Model.Builder builder = new PaymentResultHeader.Model.Builder();
        builder.setIconImage(model.getIconId() == 0 ? R.drawable.px_icon_product : model.getIconId());
        builder.setIconUrl(model.getImageUrl());
        final PaymentResultType type = PaymentResultType.from(model.getCongratsType());
        return builder
            .setBackground(type.resColor)
            .setBadgeImage(type.badge)
            .setTitle(new GenericLocalized(model.getTitle(), 0))
            .setLabel(new GenericLocalized(model.getSubtitle(), type.message))
            .build();
    }
}

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/features/business_result/BusinessPaymentResultMapper.kt
package com.mercadopago.android.px.internal.features.business_result

import com.mercadopago.android.px.R
import com.mercadopago.android.px.internal.features.payment_congrats.model.CongratsViewModelMapper
import com.mercadopago.android.px.internal.features.payment_congrats.model.PaymentCongratsModel
import com.mercadopago.android.px.internal.features.payment_congrats.model.PaymentCongratsModel.CongratsType
import com.mercadopago.android.px.internal.features.payment_congrats.model.PaymentCongratsResponse
import com.mercadopago.android.px.internal.features.payment_result.CongratsAutoReturn
import com.mercadopago.android.px.internal.features.payment_result.presentation.PaymentResultButton
import com.mercadopago.android.px.internal.features.payment_result.presentation.PaymentResultFooter
import com.mercadopago.android.px.internal.view.PaymentResultBody
import com.mercadopago.android.px.internal.view.PaymentResultHeader
import com.mercadopago.android.px.internal.view.PaymentResultMethod
import com.mercadopago.android.px.internal.viewmodel.GenericLocalized
import com.mercadopago.android.px.internal.viewmodel.PaymentResultType
import com.mercadopago.android.px.internal.mappers.Mapper
import com.mercadopago.android.px.tracking.internal.MPTracker
import java.util.*

internal class BusinessPaymentResultMapper(private val tracker: MPTracker) :
    Mapper<PaymentCongratsModel, BusinessPaymentResultViewModel>() {

    override fun map(model: PaymentCongratsModel): BusinessPaymentResultViewModel {
        return BusinessPaymentResultViewModel(
            getHeaderModel(model),
            getBodyModel(model),
            getFooterModel(model),
            getAutoReturnModel(model.paymentCongratsResponse?.autoReturn)
        )
    }

    private fun getBodyModel(model: PaymentCongratsModel): PaymentResultBody.Model {
        val methodModels: MutableList<PaymentResultMethod.Model> = ArrayList()
        if (model.shouldShowPaymentMethod == true) {
            for (paymentInfo in model.paymentsInfo) {
                methodModels.add(PaymentResultMethod.Model.with(paymentInfo,
                    model.statementDescription))
            }
        }
        val type = model.congratsType
        return PaymentResultBody.Model.Builder()
            .setMethodModels(methodModels)
            .apply { model.paymentCongratsResponse?.let {
                setCongratsViewModel(CongratsViewModelMapper(BusinessPaymentResultTracker(tracker)).map(it))
            }}
            .apply {
                if (type == CongratsType.APPROVED && model.shouldShowReceipt == true) setReceiptId(model.paymentId.toString())
            }
            .setHelp(model.help)
            .setStatement(model.statementDescription)
            .setTopFragment(model.topFragment)
            .setBottomFragment(model.bottomFragment)
            .setImportantFragment(model.importantFragment)
            .build()
    }

    private fun getHeaderModel(model: PaymentCongratsModel): PaymentResultHeader.Model {
        val type = PaymentResultType.from(model.congratsType)
        return PaymentResultHeader.Model.Builder()
            .setIconUrl(model.imageUrl)
            .setIconImage(if (model.iconId == 0) R.drawable.px_icon_product else model.iconId)
            .setBackground(type.resColor)
            .setBadgeImage(type.badge)
            .setTitle(GenericLocalized(model.title, 0))
            .setLabel(GenericLocalized(model.subtitle, type.message))
            .build()
    }

    private fun getFooterModel(model: PaymentCongratsModel): PaymentResultFooter.Model {
        return PaymentResultFooter.Model(
            model.footerMainAction?.let {
                PaymentResultButton(PaymentResultButton.Type.LOUD, it)
            },
            model.footerSecondaryAction?.let {
                PaymentResultButton(PaymentResultButton.Type.TRANSPARENT, it)
            }
        )
    }

    private fun getAutoReturnModel(autoReturn: PaymentCongratsResponse.AutoReturn?): CongratsAutoReturn.Model? {
        return autoReturn?.let {
            CongratsAutoReturn.Model(it.label, it.seconds)
        }
    }
}
